README.md
# simply_responsive

![simply_responsive](https://github.com/moseskarunia/simply-responsive/workflows/simply_responsive/badge.svg) [![codecov](https://codecov.io/gh/moseskarunia/simply-responsive/branch/master/graph/badge.svg?token=LBQMDK2MUU)](https://codecov.io/gh/moseskarunia/simply-responsive) [![pub package](https://img.shields.io/pub/v/simply_responsive.svg)](https://pub.dev/packages/simply_responsive)

<a href="https://www.buymeacoffee.com/moseskarunia" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" height="50" style="height: 6px !important;width: 22px !important;" ></a>

A mobile-first widget to help us build a responsive app across mobile, tablet, and desktop with ease. Unlike the CSS 12 grid system, The design of simply_responsive is based on 3-columns design of a desktop website. This has an advantage of being simpler to use than the CSS 12 grid system, at the cost of reduced possible flexibilities. I try to make this library as tested as possible.

And, since it's responsive, it'll be no need for you to lock the orientation of your app to portrait only!

## Upcoming Timeline
- Simply Responsive Navigation Column
- Simply Responsive Right Column and App Bar (auto show drawer & endDrawer)

## Getting Started
There are 3 kind of screen sizes breakpoints with different configuration:
1. < 600px: Mobile Screen
  - Only center column
2. < 768px: Tablet Screen
  - All possible mobile screen configuration
  - Left and center column
  - Center and right column
  - Left and right column
3. \>=768px: Desktop
  - All possible tablet screen configuration
  - All 3 columns (left, center, right)

## How to Use

```dart
SimplyResponsiveBody(
  LayoutConfig.build(MediaQuery.of(context).size.width),
  leftChild: ListView(
    children: [
      Text('Home'),
      Text('Profile'),
      Text('Sign Out'),
    ]
  ),
  centerChild: ListView(
    children: [
      Text('Lorem Ipsum'),
      Text('Lorem ipsum dolor sit amet')
    ]
  ),
  rightChild: ListView(
    children: [
      Text('Tap to like')
      Text('Tap to dislike')
    ]
  )
)
```

Each `leftChild`, `centerChild`, and `rightChild` will always have finite width. 

For more info, you can read the dart doc I wrote directly in the library code.

CHANGELOG.md
## [0.0.3] - 16 December 2020

* Update Readme

## [0.0.2] - 8 December 2020

* [BUG] Set top and bottom of positioned left and right column to 0.
* `edgePadding` default value is now 0.
* Enable 2 columns mode in mobile size.

## [0.0.1] - 8 December 2020

* LayoutConfig and SimplyResponsiveBody. The foundation of responsive scaffold in later releases. You can use them with minimal dependencies.


test/src/layout_config_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:simply_responsive/src/layout_config.dart';

void main() {
  group('should return mobile configuration', () {
    test('only center', () {
      final config = LayoutConfig.build(
        500,
        edgePadding: 16,
      );

      expect(config.centerLeftHorizontalPadding, 16);
      expect(config.centerRightHorizontalPadding, 16);
      expect(config.drawerWidth, 0.75 * 500);
      expect(config.edgePadding, 16);
      expect(config.endDrawerWidth, 0.75 * 500);
      expect(config.isLeftColumnVisible, false);
      expect(config.isRightColumnVisible, false);
      expect(config.leftColumnWidth, 0);
      expect(config.maxVisibleActionButtons, 2);
      expect(config.rightColumnWidth, 0);
      expect(config.screenWidth, 500);
      expect(config.hasLeftColumn, false);
      expect(config.hasRightColumn, false);
    });
    test('left right (2 column) mode', () {
      final config = LayoutConfig.build(
        500,
        centerFlex: 0,
        edgePadding: 16,
      );

      expect(config.centerLeftHorizontalPadding, 250);
      expect(config.centerRightHorizontalPadding, 250);
      expect(config.drawerWidth, 0.75 * 500);
      expect(config.edgePadding, 16);
      expect(config.endDrawerWidth, 0.75 * 500);
      expect(config.isLeftColumnVisible, true);
      expect(config.isRightColumnVisible, true);
      expect(config.leftColumnWidth, 250 - 16);
      expect(config.maxVisibleActionButtons, 2);
      expect(config.rightColumnWidth, 250 - 16);
      expect(config.screenWidth, 500);
      expect(config.hasLeftColumn, true);
      expect(config.hasRightColumn, true);
    });
  });

  group('should return tablet configuration', () {
    test('with left column', () {
      final config = LayoutConfig.build(
        750,
        edgePadding: 16,
      );

      final leftWidth = (1 * 750 / 3).floor();

      expect(config.centerLeftHorizontalPadding, leftWidth);
      expect(config.centerRightHorizontalPadding, 16);
      expect(config.drawerWidth, 0.5 * 750);
      expect(config.edgePadding, 16);
      expect(config.endDrawerWidth, 0.5 * 750);
      expect(config.isLeftColumnVisible, true);
      expect(config.isRightColumnVisible, false);
      expect(config.leftColumnWidth, leftWidth - 16);
      expect(config.maxVisibleActionButtons, 3);
      expect(config.rightColumnWidth, 0);
      expect(config.screenWidth, 750);
      expect(config.hasLeftColumn, true);
      expect(config.hasRightColumn, false);
    });
    test('with right column instead of left', () {
      final config = LayoutConfig.build(
        750,
        leftFlex: 0,
        edgePadding: 16,
      );

      final rightWidth = (1 * 750 / 3).floor();

      expect(config.centerLeftHorizontalPadding, 16);
      expect(config.centerRightHorizontalPadding, rightWidth);
      expect(config.drawerWidth, 0.5 * 750);
      expect(config.edgePadding, 16);
      expect(config.endDrawerWidth, 0.5 * 750);
      expect(config.isLeftColumnVisible, false);
      expect(config.isRightColumnVisible, true);
      expect(config.leftColumnWidth, 0);
      expect(config.rightColumnWidth, rightWidth - 16);
      expect(config.screenWidth, 750);
      expect(config.hasLeftColumn, false);
      expect(config.hasRightColumn, true);
    });
    test('with center column only', () {
      final config = LayoutConfig.build(
        750,
        leftFlex: 0,
        rightFlex: 0,
        edgePadding: 16,
      );

      expect(config.centerLeftHorizontalPadding, 93);
      expect(config.centerRightHorizontalPadding, 93);
      expect(config.drawerWidth, 0.5 * 750);
      expect(config.edgePadding, 93);
      expect(config.endDrawerWidth, 0.5 * 750);
      expect(config.isLeftColumnVisible, false);
      expect(config.isRightColumnVisible, false);
      expect(config.leftColumnWidth, 0);
      expect(config.maxVisibleActionButtons, 3);
      expect(config.rightColumnWidth, 0);
      expect(config.screenWidth, 750);
      expect(config.hasLeftColumn, false);
      expect(config.hasRightColumn, false);
    });

    test('with left and right column only', () {
      final config = LayoutConfig.build(
        750,
        leftFlex: 1,
        rightFlex: 1,
        centerFlex: 0,
        edgePadding: 16,
      );

      expect(config.centerLeftHorizontalPadding, 375);
      expect(config.centerRightHorizontalPadding, 375);
      expect(config.drawerWidth, 0.5 * 750);
      expect(config.edgePadding, 0);
      expect(config.endDrawerWidth, 0.5 * 750);
      expect(config.isLeftColumnVisible, true);
      expect(config.isRightColumnVisible, true);
      expect(config.leftColumnWidth, 375);
      expect(config.maxVisibleActionButtons, 3);
      expect(config.rightColumnWidth, 375);
      expect(config.screenWidth, 750);
      expect(config.hasLeftColumn, true);
      expect(config.hasRightColumn, true);
    });
  });

  group('should return desktop configuration', () {
    test('with both side columns', () {
      final config = LayoutConfig.build(
        1000,
        edgePadding: 16,
      );

      expect(config.centerLeftHorizontalPadding, 250);
      expect(config.centerRightHorizontalPadding, 250);
      expect(config.drawerWidth, 0.25 * 1000);
      expect(config.edgePadding, 16);
      expect(config.endDrawerWidth, 0.25 * 1000);
      expect(config.isLeftColumnVisible, true);
      expect(config.isRightColumnVisible, true);
      expect(config.leftColumnWidth, 250 - 16);
      expect(config.maxVisibleActionButtons, 5);
      expect(config.rightColumnWidth, 250 - 16);
      expect(config.screenWidth, 1000);
      expect(config.hasLeftColumn, true);
      expect(config.hasRightColumn, true);
    });
    test('with left column', () {
      final config = LayoutConfig.build(
        1000,
        rightFlex: 0,
        edgePadding: 16,
      );
      final leftWidth = (1000 / 3).floor();

      expect(config.centerLeftHorizontalPadding, leftWidth);
      expect(config.centerRightHorizontalPadding, 16);
      expect(config.drawerWidth, 0.25 * 1000);
      expect(config.edgePadding, 16);
      expect(config.endDrawerWidth, 0.25 * 1000);
      expect(config.isLeftColumnVisible, true);
      expect(config.isRightColumnVisible, false);
      expect(config.leftColumnWidth, leftWidth - 16);
      expect(config.maxVisibleActionButtons, 5);
      expect(config.rightColumnWidth, 0);
      expect(config.screenWidth, 1000);
      expect(config.hasLeftColumn, true);
      expect(config.hasRightColumn, false);
    });
    test('with right column', () {
      final config = LayoutConfig.build(
        1000,
        leftFlex: 0,
        edgePadding: 16,
      );
      final rightWidth = (1000 / 3).floor();

      expect(config.centerLeftHorizontalPadding, 16);
      expect(config.centerRightHorizontalPadding, rightWidth);
      expect(config.drawerWidth, 0.25 * 1000);
      expect(config.edgePadding, 16);
      expect(config.endDrawerWidth, 0.25 * 1000);
      expect(config.isLeftColumnVisible, false);
      expect(config.isRightColumnVisible, true);
      expect(config.leftColumnWidth, 0);
      expect(config.maxVisibleActionButtons, 5);
      expect(config.rightColumnWidth, rightWidth - 16);
      expect(config.screenWidth, 1000);
      expect(config.hasLeftColumn, false);
      expect(config.hasRightColumn, true);
    });
    test('with center column only', () {
      final config = LayoutConfig.build(
        1000,
        leftFlex: 0,
        rightFlex: 0,
        edgePadding: 16,
      );
      final sidePadding = (1000 * 0.25 / 2).floor();

      expect(config.centerLeftHorizontalPadding, sidePadding);
      expect(config.centerRightHorizontalPadding, sidePadding);
      expect(config.drawerWidth, 0.25 * 1000);
      expect(config.edgePadding, sidePadding);
      expect(config.endDrawerWidth, 0.25 * 1000);
      expect(config.isLeftColumnVisible, false);
      expect(config.isRightColumnVisible, false);
      expect(config.leftColumnWidth, 0);
      expect(config.maxVisibleActionButtons, 5);
      expect(config.rightColumnWidth, 0);
      expect(config.screenWidth, 1000);
      expect(config.hasLeftColumn, false);
      expect(config.hasRightColumn, false);
    });

    test('with left and right column only', () {
      final config = LayoutConfig.build(
        1200,
        leftFlex: 2,
        rightFlex: 1,
        centerFlex: 0,
        edgePadding: 16,
      );

      expect(config.centerLeftHorizontalPadding, 800);
      expect(config.centerRightHorizontalPadding, 400);
      expect(config.drawerWidth, 0.25 * 1200);
      expect(config.edgePadding, 0);
      expect(config.endDrawerWidth, 0.25 * 1200);
      expect(config.isLeftColumnVisible, true);
      expect(config.isRightColumnVisible, true);
      expect(config.leftColumnWidth, 800);
      expect(config.maxVisibleActionButtons, 5);
      expect(config.rightColumnWidth, 400);
      expect(config.screenWidth, 1200);
      expect(config.hasLeftColumn, true);
      expect(config.hasRightColumn, true);
    });
  });
}


test/src/arrange_scaffold_child_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:simply_responsive/src/arrange_scaffold_child.dart';
import 'package:simply_responsive/src/layout_config.dart';

void main() {
  final arrangeScaffoldChild = const ArrangeScaffoldChild();

  group('should return only list view in center column', () {
    testWidgets('when no leftChildren nor rightChildren', (tester) async {
      final configFixture = LayoutConfig.build(500);
      final widget = arrangeScaffoldChild(
        configFixture,
        centerChildren: <Widget>[
          Text('1'),
          Text('2'),
        ],
        centerSeparator: SizedBox(height: 16),
      );

      await tester.pumpWidget(MaterialApp(home: Scaffold(body: widget)));

      expect(widget, isA<ListView>());
      expect(find.byType(ListView), findsOneWidget);
      expect(
        find.ancestor(of: find.text('1'), matching: find.byType(ListView)),
        findsOneWidget,
      );
      expect(
        find.ancestor(of: find.text('2'), matching: find.byType(ListView)),
        findsOneWidget,
      );
      expect(
        tester.widget<ListView>(find.byType(ListView)).padding,
        EdgeInsets.fromLTRB(
          configFixture.centerLeftHorizontalPadding.toDouble(),
          16,
          configFixture.centerRightHorizontalPadding.toDouble(),
          0,
        ),
      );
      expect(
        find.descendant(
          of: find.byKey(Key('centerColumn')),
          matching: find.byWidgetPredicate(
            (w) => w is SizedBox && w.height == 16,
          ),
        ),
        findsNWidgets(1),
      );
      expect(find.byKey(Key('leftColumn')), findsNothing);
      expect(find.byKey(Key('rightColumn')), findsNothing);
    });
    testWidgets('when leftColumnWidth and rightColumnWidth equals 0',
        (tester) async {
      final configFixture = LayoutConfig.build(750, leftFlex: 0, rightFlex: 0);
      final widget = arrangeScaffoldChild(
        configFixture,
        centerChildren: <Widget>[Text('1'), Text('2')],
        leftChildren: <Widget>[Text('a'), Text('b')],
        rightChildren: <Widget>[Text('ax'), Text('bx')],
        centerSeparator: SizedBox(height: 16),
      );
      await tester.pumpWidget(MaterialApp(home: Scaffold(body: widget)));
      expect(widget, isA<ListView>());
      expect(find.byType(ListView), findsOneWidget);
      expect(
        find.ancestor(of: find.text('1'), matching: find.byType(ListView)),
        findsOneWidget,
      );
      expect(
        find.ancestor(of: find.text('2'), matching: find.byType(ListView)),
        findsOneWidget,
      );
      expect(
        tester.widget<ListView>(find.byType(ListView)).padding,
        EdgeInsets.fromLTRB(
          configFixture.centerLeftHorizontalPadding.toDouble(),
          16,
          configFixture.centerRightHorizontalPadding.toDouble(),
          0,
        ),
      );
      expect(
        find.descendant(
          of: find.byKey(Key('centerColumn')),
          matching: find.byWidgetPredicate(
            (w) => w is SizedBox && w.height == 16,
          ),
        ),
        findsNWidgets(1),
      );
      expect(find.byKey(Key('leftColumn')), findsNothing);
      expect(find.byKey(Key('rightColumn')), findsNothing);
    });
  });
}


test/src/simply_responsive_body_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:simply_responsive/src/layout_config.dart';
import 'package:simply_responsive/src/simply_responsive_body.dart';

void main() {
  group('should return with all three columns:', () {
    final config = LayoutConfig.build(1000);
    SimplyResponsiveBody body;
    setUp(() {
      /// Otherwise defaults to 800x600
      WidgetsBinding.instance.renderView.configuration =
          new TestViewConfiguration(size: const Size(1024, 768));

      body = SimplyResponsiveBody(
        config,
        centerChild: Text('the center'),
        leftChild: Text('the left'),
        rightChild: Text('the right'),
      );
    });

    test('should throw AssertionError if no layout config', () async {
      await expectLater(
        () => SimplyResponsiveBody(
          null,
          centerChild: Text('the center'),
          leftChild: Text('the left'),
          rightChild: Text('the right'),
        ),
        throwsA(isA<AssertionError>()),
      );
    });

    testWidgets('center column is correct', (tester) async {
      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));

      final centerWidget = find.byKey(Key('centerColumn'));
      expect(tester.getSize(centerWidget).width, config.screenWidth);
      expect(
        find.ancestor(of: find.text('the center'), matching: centerWidget),
        findsOneWidget,
      );

      final centerPositioned = find.ancestor(
        of: centerWidget,
        matching: find.byType(Positioned),
      );
      expect(centerPositioned, findsOneWidget);
    });

    testWidgets('left column is correct', (tester) async {
      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));

      final leftWidget = find.byKey(Key('leftColumn'));
      expect(tester.getSize(leftWidget).width, config.leftColumnWidth);

      final leftPositioned = find.ancestor(
        of: leftWidget,
        matching: find.byType(Positioned),
      );
      expect(leftPositioned, findsOneWidget);
      expect(
        tester.widget<Positioned>(leftPositioned).left,
        config.edgePadding,
      );
      expect(
        find.ancestor(
          of: leftPositioned,
          matching: find.byType(Stack),
        ),
        findsOneWidget,
      );
    });

    testWidgets('right column is correct', (tester) async {
      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));

      final rightWidget = find.byKey(Key('rightColumn'));
      expect(tester.getSize(rightWidget).width, config.rightColumnWidth);

      final rightPositioned = find.ancestor(
        of: find.text('the right'),
        matching: find.byType(Positioned),
      );
      expect(rightPositioned, findsOneWidget);
      expect(
        tester.widget<Positioned>(rightPositioned).right,
        config.edgePadding,
      );
      expect(
        find.ancestor(
          of: rightPositioned,
          matching: find.byType(Stack),
        ),
        findsOneWidget,
      );
    });

    testWidgets('outer container', (tester) async {
      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));

      final rightWidget = find.byType(Stack);
      final outerContainer = find.ancestor(
        of: rightWidget,
        matching: find.byType(Container),
      );

      expect(tester.getSize(outerContainer).width, config.screenWidth);
    });
  });

  group('should return only the center column', () {
    void performAssertion(WidgetTester tester, LayoutConfig config) {
      final centerWidget = find.byKey(Key('centerColumn'));
      final leftWidget = find.byKey(Key('leftColumn'));
      final rightWidget = find.byKey(Key('rightColumn'));

      expect(centerWidget, findsOneWidget);
      expect(leftWidget, findsNothing);
      expect(rightWidget, findsNothing);

      expect(tester.getSize(centerWidget).width, config.screenWidth);

      /// When only center column, the direct ancestor should be a scaffold
      expect(
        find.ancestor(of: centerWidget, matching: find.byType(Scaffold)),
        findsOneWidget,
      );
    }

    testWidgets('if both !hasLeftColumn & !hasRightColumn', (tester) async {
      final config = LayoutConfig(
        screenWidth: 500,
        leftColumnWidth: 0,
        rightColumnWidth: 0,
        isLeftColumnVisible: true,
        isRightColumnVisible: true,
      );
      final body = SimplyResponsiveBody(
        config,
        centerChild: Text('the center'),
        leftChild: Text('the left'),
        rightChild: Text('the right'),
      );

      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));

      performAssertion(tester, config);
    });

    testWidgets('if left & right child is null', (tester) async {
      final config = LayoutConfig(
        screenWidth: 500,
        leftColumnWidth: 100,
        rightColumnWidth: 100,
        isLeftColumnVisible: true,
        isRightColumnVisible: true,
      );
      final body = SimplyResponsiveBody(
        config,
        centerChild: Text('the center'),
        leftChild: null,
        rightChild: null,
      );

      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));

      performAssertion(tester, config);
    });
  });

  group('should return left and center only', () {
    void performAssertion(WidgetTester tester, LayoutConfig config) {
      final centerWidget = find.byKey(Key('centerColumn'));
      final leftWidget = find.byKey(Key('leftColumn'));
      final rightWidget = find.byKey(Key('rightColumn'));
      expect(centerWidget, findsOneWidget);
      expect(leftWidget, findsOneWidget);
      expect(rightWidget, findsNothing);

      expect(tester.getSize(centerWidget).width, config.screenWidth);
      expect(tester.getSize(leftWidget).width, config.leftColumnWidth);

      final leftPositioned = find.ancestor(
        of: leftWidget,
        matching: find.byType(Positioned),
      );
      expect(leftPositioned, findsOneWidget);
      expect(
        tester.widget<Positioned>(leftPositioned).left,
        config.edgePadding,
      );
      expect(
        find.ancestor(of: leftPositioned, matching: find.byType(Stack)),
        findsOneWidget,
      );
      final centerPositioned = find.ancestor(
        of: centerWidget,
        matching: find.byType(Positioned),
      );
      expect(centerPositioned, findsOneWidget);
      expect(find.byType(Positioned), findsNWidgets(2));

      expect(
        find.ancestor(of: leftPositioned, matching: find.byType(Stack)),
        findsOneWidget,
      );
      expect(
        find.ancestor(of: centerPositioned, matching: find.byType(Stack)),
        findsOneWidget,
      );
    }

    testWidgets('if hasLeftColumn && !hasRightColumn', (tester) async {
      final config = LayoutConfig(
        screenWidth: 500,
        leftColumnWidth: 100,
        rightColumnWidth: 0,
        isLeftColumnVisible: true,
        isRightColumnVisible: false,
        edgePadding: 16,
      );

      final body = SimplyResponsiveBody(
        config,
        centerChild: Text('the center'),
        leftChild: Text('the left'),
        rightChild: Text('the right'),
      );

      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));
      performAssertion(tester, config);
    });

    testWidgets('if hasLeftColumn && no rightChild', (tester) async {
      final config = LayoutConfig(
        screenWidth: 500,
        leftColumnWidth: 100,
        rightColumnWidth: 100,
        isLeftColumnVisible: true,
        isRightColumnVisible: true,
        edgePadding: 16,
      );

      final body = SimplyResponsiveBody(
        config,
        centerChild: Text('the center'),
        leftChild: Text('the left'),
        rightChild: null,
      );

      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));
      performAssertion(tester, config);
    });
  });

  group('should return right and center only', () {
    void performAssertion(WidgetTester tester, LayoutConfig config) {
      final centerWidget = find.byKey(Key('centerColumn'));
      final leftWidget = find.byKey(Key('leftColumn'));
      final rightWidget = find.byKey(Key('rightColumn'));
      expect(centerWidget, findsOneWidget);
      expect(leftWidget, findsNothing);
      expect(rightWidget, findsOneWidget);

      expect(tester.getSize(centerWidget).width, config.screenWidth);
      expect(tester.getSize(rightWidget).width, config.rightColumnWidth);

      final rightPositioned = find.ancestor(
        of: rightWidget,
        matching: find.byType(Positioned),
      );
      expect(rightPositioned, findsOneWidget);
      expect(
        tester.widget<Positioned>(rightPositioned).right,
        config.edgePadding,
      );
      expect(
        find.ancestor(of: rightPositioned, matching: find.byType(Stack)),
        findsOneWidget,
      );
      final centerPositioned = find.ancestor(
        of: centerWidget,
        matching: find.byType(Positioned),
      );
      expect(centerPositioned, findsOneWidget);
      expect(find.byType(Positioned), findsNWidgets(2));

      expect(
        find.ancestor(of: rightPositioned, matching: find.byType(Stack)),
        findsOneWidget,
      );
      expect(
        find.ancestor(of: centerPositioned, matching: find.byType(Stack)),
        findsOneWidget,
      );
    }

    testWidgets('if hasRightColumn && !hasLeftColumn', (tester) async {
      final config = LayoutConfig(
        screenWidth: 500,
        leftColumnWidth: 0,
        rightColumnWidth: 100,
        isLeftColumnVisible: false,
        isRightColumnVisible: true,
        edgePadding: 16,
      );

      final body = SimplyResponsiveBody(
        config,
        centerChild: Text('the center'),
        leftChild: Text('the left'),
        rightChild: Text('the right'),
      );

      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));
      performAssertion(tester, config);
    });

    testWidgets('if hasRightColumn && no leftChild', (tester) async {
      final config = LayoutConfig(
        screenWidth: 500,
        leftColumnWidth: 100,
        rightColumnWidth: 100,
        isLeftColumnVisible: true,
        isRightColumnVisible: true,
        edgePadding: 16,
      );

      final body = SimplyResponsiveBody(
        config,
        centerChild: Text('the center'),
        leftChild: null,
        rightChild: Text('the right'),
      );

      await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));
      performAssertion(tester, config);
    });
  });

  testWidgets('should build 2 columns mode', (tester) async {
    final config = LayoutConfig(
      screenWidth: 500,
      leftColumnWidth: 300,
      rightColumnWidth: 200,
      isLeftColumnVisible: true,
      isRightColumnVisible: true,
      edgePadding: 16,
    );

    final body = SimplyResponsiveBody(
      config,
      centerChild: null,
      leftChild: Text('the left'),
      rightChild: Text('the right'),
    );
    await tester.pumpWidget(MaterialApp(home: Scaffold(body: body)));

    final centerWidget = find.byKey(Key('centerColumn'));
    final leftWidget = find.byKey(Key('leftColumn'));
    final rightWidget = find.byKey(Key('rightColumn'));
    expect(centerWidget, findsNothing);
    expect(leftWidget, findsOneWidget);
    expect(rightWidget, findsOneWidget);

    expect(tester.getSize(leftWidget).width, config.leftColumnWidth);
    expect(tester.getSize(rightWidget).width, config.rightColumnWidth);

    final row1 = find.ancestor(of: leftWidget, matching: find.byType(Row));
    final row2 = find.ancestor(of: rightWidget, matching: find.byType(Row));
    expect(row1, findsOneWidget);
    expect(row2, findsOneWidget);

    expect(tester.widget<Row>(row1).children.length, 2);

    final outerContainer =
        find.ancestor(of: row1, matching: find.byType(Container));

    expect(tester.getSize(outerContainer).width, config.screenWidth);
  });
}


lib/src/layout_config.dart
import 'dart:math';

import 'package:flutter/widgets.dart';

/// Build informations screen sizes.
/// This object is recalculated on each build call.
class LayoutConfig {
  final double screenWidth;

  /// Width of the left column.
  final double leftColumnWidth;

  /// Width of the drawer.
  final double drawerWidth;

  /// Width of the right column.
  final double rightColumnWidth;

  /// Width of the endDrawer.
  final double endDrawerWidth;

  /// When false, the left and right column won't be visible,
  /// no matter their respective widths.
  final bool isLeftColumnVisible, isRightColumnVisible;

  /// Edge padding to make the center scrollable from the edge of the screen.
  /// The same value will be substracted from both left and right column width
  /// (if any). Default value = 16.
  ///
  /// This value is used as a left and right value of Positioned widget
  /// of left and right column.
  final double edgePadding;

  /// Maximum allowed action buttons at the top right of the scaffold.
  /// Default 2
  final int maxVisibleActionButtons;

  /// Use both [centerLeftHorizontalPadding] and [centerRightHorizontalPadding]
  /// as your scrollable widget's padding to make it scrollable from the edge.
  double get centerLeftHorizontalPadding => edgePadding + leftColumnWidth;
  double get centerRightHorizontalPadding => edgePadding + rightColumnWidth;

  /// A computed property which combines [leftColumnWidth] &
  /// [isLeftColumnVisible] value.
  bool get hasLeftColumn => leftColumnWidth > 0 && isLeftColumnVisible;

  /// A computed property which combines [rightColumnWidth] &
  /// [isRightColumnVisible] value.
  bool get hasRightColumn => rightColumnWidth > 0 && isRightColumnVisible;

  const LayoutConfig({
    this.drawerWidth = 0,
    this.edgePadding = 16,
    this.endDrawerWidth = 0,
    this.isLeftColumnVisible = false,
    this.isRightColumnVisible = false,
    this.leftColumnWidth = 0,
    this.maxVisibleActionButtons = 2,
    this.rightColumnWidth = 0,
    @required this.screenWidth,
  });

  /// Build layout config from [screenWidth]. The center will be calculated
  /// with ceiling rounding first.
  ///
  /// There's normal mode and 2 columns only mode, which splits the screen in 2.
  ///
  /// * [drawerToWidthRatio] ratio of drawer to the total width of the screen.
  ///   When null, the default value is: 0.75 on mobile, 0.5 on tablet, and
  ///   0.25 on desktop
  /// * [endDrawerRatio] ratio of drawer to the total width of the screen.
  ///   When null, the default value is: 0.75 on mobile, 0.5 on tablet, and
  ///   0.25 on desktop
  /// * [edgePadding] The space on the left and right of the screen to make
  ///   the center scrollable from the edge. The same amount will automatically.
  ///   substracted from left and right width (if not already 0).
  /// * [leftFlex] will be ignored in mobile size. To make the left column
  ///   hidden, set to 0
  /// * [maxCenterToScreenRatioWhenNoSideColumn] Max center column width ratio
  ///   to screenWidth. Will be ignored unless left and right flex is 0. The
  ///   purpose of this property is to improve page readability by reducing
  ///   user's horizontal eye movement
  /// * [maxVisibleActionButtons] maximum allowed action buttons on the top
  ///   right of the app bar. Default value, 2 on mobile, 3 on tablet,
  ///   and 5 on desktop and wider.
  /// * [rightFlex] will be ignored in mobile size. Will be ignored in tablet.
  ///   when [leftFlex] is not 0. To make the right column hidden, set to 0.

  factory LayoutConfig.build(
    double screenWidth, {
    int centerFlex = 2,
    double drawerToScreenWidthRatio,
    double edgePadding = 0,
    double endDrawerToScreenWidthRatio,
    int leftFlex = 1,
    int maxVisibleActionButtons,
    int rightFlex = 1,
    double maxCenterToScreenRatioWhenNoSideColumn = 0.75,
  }) {
    int calculatedMaxVisibleActionButtons;
    double calculatedDrawerRatio, calculatedEndDrawerRatio;
    if (screenWidth < 600) {
      // Mobile
      calculatedMaxVisibleActionButtons = maxVisibleActionButtons ?? 2;
      calculatedDrawerRatio = drawerToScreenWidthRatio ?? 0.75;
      calculatedEndDrawerRatio = endDrawerToScreenWidthRatio ?? 0.75;

      double calculatedLeftWidth = 0, calculatedRightWidth = 0;
      bool calculatedIsLeftColumnVisible = false,
          calculatedIsRightColumnVisible = false;

      if (centerFlex == 0) {
        int totalFlex = leftFlex + rightFlex;
        calculatedLeftWidth =
            max((leftFlex * screenWidth / totalFlex).floorToDouble(), 0);
        calculatedRightWidth =
            max((rightFlex * screenWidth / totalFlex).floorToDouble(), 0);

        if (leftFlex > 0) {
          calculatedIsLeftColumnVisible = true;
        }
        if (rightFlex > 0) {
          calculatedIsRightColumnVisible = true;
        }
      }

      return LayoutConfig(
        drawerWidth: calculatedDrawerRatio * screenWidth,
        endDrawerWidth: calculatedEndDrawerRatio * screenWidth,
        isLeftColumnVisible: calculatedIsLeftColumnVisible,
        isRightColumnVisible: calculatedIsRightColumnVisible,
        leftColumnWidth: max(calculatedLeftWidth - edgePadding, 0),
        maxVisibleActionButtons: calculatedMaxVisibleActionButtons,
        rightColumnWidth: max(calculatedRightWidth - edgePadding, 0),
        screenWidth: screenWidth,
        edgePadding: edgePadding,
      );
    }

    int totalFlex;
    double leftWidth, rightWidth;
    double calculatedEdgePadding;

    if (screenWidth < 768) {
      // Tablet
      calculatedMaxVisibleActionButtons = maxVisibleActionButtons ?? 3;
      calculatedDrawerRatio = drawerToScreenWidthRatio ?? 0.5;
      calculatedEndDrawerRatio = endDrawerToScreenWidthRatio ?? 0.5;

      if (centerFlex > 0) {
        totalFlex =
            leftFlex > 0 ? centerFlex + leftFlex : centerFlex + rightFlex;
        leftWidth =
            max((leftFlex * screenWidth / totalFlex).floorToDouble(), 0);
        rightWidth = leftFlex == 0
            ? max((rightFlex * screenWidth / totalFlex).floorToDouble(), 0)
            : 0;
        calculatedEdgePadding = edgePadding;
      } else {
        totalFlex = leftFlex + rightFlex;
        leftWidth =
            max((leftFlex * screenWidth / totalFlex).floorToDouble(), 0);
        rightWidth =
            max((rightFlex * screenWidth / totalFlex).floorToDouble(), 0);
        calculatedEdgePadding = 0;
      }

      if (maxCenterToScreenRatioWhenNoSideColumn < 1 &&
          leftFlex == 0 &&
          rightFlex == 0) {
        calculatedEdgePadding =
            ((1.0 - maxCenterToScreenRatioWhenNoSideColumn) * screenWidth / 2)
                .floorToDouble();
      }

      return LayoutConfig(
        drawerWidth: (calculatedDrawerRatio * screenWidth).floorToDouble(),
        endDrawerWidth:
            (calculatedEndDrawerRatio * screenWidth).floorToDouble(),
        isLeftColumnVisible: leftFlex > 0,
        isRightColumnVisible: (leftFlex == 0 && rightFlex > 0) ||
            centerFlex == 0 && rightFlex > 0,
        leftColumnWidth: max(leftWidth - calculatedEdgePadding, 0),
        maxVisibleActionButtons: calculatedMaxVisibleActionButtons,
        rightColumnWidth: max(rightWidth - calculatedEdgePadding, 0),
        screenWidth: screenWidth,
        edgePadding: calculatedEdgePadding,
      );
    }
    // Desktop or wider

    calculatedMaxVisibleActionButtons = maxVisibleActionButtons ?? 5;
    calculatedDrawerRatio = drawerToScreenWidthRatio ?? 0.25;
    calculatedEndDrawerRatio = endDrawerToScreenWidthRatio ?? 0.25;

    totalFlex = centerFlex + leftFlex + rightFlex;

    leftWidth = max((leftFlex * screenWidth / totalFlex).floorToDouble(), 0);
    rightWidth = max((rightFlex * screenWidth / totalFlex).floorToDouble(), 0);

    calculatedEdgePadding = edgePadding;

    if (centerFlex == 0 && leftFlex > 0 && rightFlex > 0) {
      calculatedEdgePadding = 0;
    }

    if (maxCenterToScreenRatioWhenNoSideColumn < 1 &&
        leftFlex == 0 &&
        rightFlex == 0) {
      calculatedEdgePadding =
          ((1.0 - maxCenterToScreenRatioWhenNoSideColumn) * screenWidth / 2)
              .floorToDouble();
    }

    return LayoutConfig(
      drawerWidth: calculatedDrawerRatio * screenWidth,
      endDrawerWidth: calculatedEndDrawerRatio * screenWidth,
      isLeftColumnVisible: leftFlex > 0,
      isRightColumnVisible: rightFlex > 0,
      leftColumnWidth: max(leftWidth - calculatedEdgePadding, 0),
      maxVisibleActionButtons: calculatedMaxVisibleActionButtons,
      rightColumnWidth: max(rightWidth - calculatedEdgePadding, 0),
      screenWidth: screenWidth,
      edgePadding: calculatedEdgePadding,
    );
  }
}
